src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,174 @@
+            currentNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexLess(index);
+        Node<T> neededNode = first;
+        for (int i = 0; i < index; i++) {
+            neededNode = neededNode.next;
+        }
+        return neededNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexLess(index);
+        Node<T> setNode = findNodeByIndex(index);
+        T oldValue = setNode.value;
+        setNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexLess(index);
+        Node<T> removedNode = findNodeByIndex(index);
+        unlink(removedNode);
+        size--;
+        return removedNode.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> redundantNode = findNodeByValue(object);
+        if (redundantNode == null) {
+            return false;
+        }
+        unlink(redundantNode);
+        size--;
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private void checkIndexLess(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + " is not exist");
+        }
+    }
+
+    private void checkIndexBigger(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + " is not exist");
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> currentNode = first;
+        if (index < size / 2) {
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = last;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    private Node<T> findNodeByValue(T object) {
+        Node<T> deletedNode = first;
+        while (deletedNode != null) {
+            if (deletedNode.value == object
+                    || deletedNode != null && deletedNode.value.equals(object)) {
+                return deletedNode;
+            }
+            deletedNode = deletedNode.next;
+        }
```

Comment:
check if index < size / 2 then move from first to last, else move from last to first

Reply:
There is no any index. This method is looking for Node by value. But in another method I use looking for from both side, see findNodeByIndex() method.